pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -291,70 +274,99 @@ pub fn schedule_timeline_checkpoint_upload(
+            }
+        });
+    let remote_index = RemoteTimelineIndex::try_parse_descriptions_from_paths(conf, download_paths);
+
+    let initial_timeline_states = schedule_first_sync_tasks(&remote_index, local_timeline_files);
+
     let handle = thread::Builder::new()
-        .name("Queue based remote storage sync".to_string())
+        .name("Remote storage sync thread".to_string())
         .spawn(move || {
-            let thread_result = storage_sync_loop(
-                config,
+            storage_sync_loop(
+                runtime,
+                conf,
                 receiver,
-                remote_storage,
+                remote_index,
+                storage,
                 max_concurrent_sync,
                 max_sync_errors,
-            );
-            if let Err(e) = &thread_result {
-                error!("Failed to run storage sync thread: {:#}", e);
-            }
-            thread_result
-        })?;
-    Ok(handle)
+            )
+        })
+        .context("Failed to spawn remote storage sync thread")?;
+    Ok(SyncStartupData {
+        initial_timeline_states,
+        sync_loop_handle: Some(handle),
+    })
 }
 
 fn storage_sync_loop<
     P: std::fmt::Debug + Send + Sync + 'static,
     S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
 >(
-    config: &'static PageServerConf,
+    runtime: tokio::runtime::Runtime,
+    conf: &'static PageServerConf,
     mut receiver: UnboundedReceiver<SyncTask>,
-    remote_storage: S,
+    index: RemoteTimelineIndex,
+    storage: S,
     max_concurrent_sync: NonZeroUsize,
     max_sync_errors: NonZeroU32,
 ) -> anyhow::Result<()> {
-    let runtime = tokio::runtime::Builder::new_current_thread()
-        .enable_all()
-        .build()?;
-    let remote_timelines = runtime
-        .block_on(fetch_existing_uploads(&remote_storage))
-        .context("Failed to determine previously uploaded timelines")?;
-
-    schedule_first_tasks(config, &remote_timelines);
-
-    // placing the two variables, shared between the async loop tasks. Main reason for using `Arc` is `tokio::spawn` with its `'static` requirements.
-    let remote_timelines_and_storage = Arc::new((Mutex::new(remote_timelines), remote_storage));
-
+    let remote_assets = Arc::new((storage, RwLock::new(index)));
     while !crate::tenant_mgr::shutdown_requested() {
-        let registration_steps = runtime.block_on(loop_step(
-            config,
+        let new_timeline_states = runtime.block_on(loop_step(
+            conf,
             &mut receiver,
-            Arc::clone(&remote_timelines_and_storage),
+            Arc::clone(&remote_assets),
             max_concurrent_sync,
             max_sync_errors,
         ));
         // Batch timeline download registration to ensure that the external registration code won't block any running tasks before.
-        perform_post_timeline_sync_steps(config, registration_steps);
+        set_timeline_states(conf, new_timeline_states);
```

Comment:
Good note, indeed, now I did not consider this much due to sync code accessing it only, but, on the other hand, I plan to expose the state enums into the `Timeline` and http endpoint to display already in this PR.

I have no good answer on this case now: holding the lock over a timeline state unpredictably long (since sync tasks can fail and retry) does not sound good.

With the current approach, we might escape the majority of the issues by blocking timeline interactions before its complete download (which happens once per timeline now). 
Checkpoint uploads do not cause the state changes, the `Evicted` seem to be fine to apply at any state.
`CloudOnly` is a bit weird, but might be fine to behave the same as `Evicted` for simplicity.

Reply:
I think for now blocking is fine, we will see what we need next and drive this accordingly. IMO it is better for this blocking to be explicit, so it is harder to misuse the api

Reply:
I think you're right and I like dashmap/CAS option better than locking.
IMO, now it's tolerable at this form, since the only code, changing the timelines the sync loop, so I'd not expect any sudden changes right now.

Future PRs might definitely improve that.
Not sure if worth creating a separate issue, since looks very related to https://github.com/zenithdb/zenith/issues/971 (as the root issue)?

Reply:
From my perspective it seems unrelated to #971. This is more about api misuse resistance or some changes in the future that might break the assumptions

Reply:
Ok, I've added the comment now, so seems that we're aware of the issue in some form, just have to fix it later at some point.